pkg/k8s/watchers/cilium_endpoint_slice_subscriber_test.go

```diff
@@ -0,0 +1,497 @@
+}
+
+func epEqual(e1, e2 *types.CiliumEndpoint) bool {
+	return ((e1 == nil) == (e2 == nil)) && (e1 == nil || e1.Identity.ID == e2.Identity.ID)
+}
+
+func updateEqual(u1, u2 endpointUpdate) bool {
+	return epEqual(u1.oldEp, u2.oldEp) && epEqual(u1.newEp, u2.newEp)
+}
+
+type fakeEPWatcher struct {
+	lastUpdate endpointUpdate
+	lastDelete *types.CiliumEndpoint
+}
+
+func createFakeEPWatcher() *fakeEPWatcher {
+	return &fakeEPWatcher{}
+}
+
+func (fw *fakeEPWatcher) endpointUpdated(oldC, newC *types.CiliumEndpoint) {
+	fw.lastUpdate = endpointUpdate{oldC, newC}
+}
+
+func (fw *fakeEPWatcher) endpointDeleted(c *types.CiliumEndpoint) {
+	fw.lastDelete = c
+}
+
+func (fw *fakeEPWatcher) expectUpdate(u endpointUpdate) (string, bool) {
+	if !updateEqual(fw.lastUpdate, u) {
+		return fmt.Sprintf("Expected %s, got %s", u.toString(), fw.lastUpdate.toString()), false
+	}
+	return "", true
+}
+
+func (fw *fakeEPWatcher) expectNoDelete() (string, bool) {
+	if fw.lastDelete != nil {
+		return fmt.Sprintf("Expected no delete, got %s", epToString(fw.lastDelete)), false
+	}
+	return "", true
+}
+
+func (fw *fakeEPWatcher) expectDelete(e *types.CiliumEndpoint) (string, bool) {
+	if !epEqual(fw.lastDelete, e) {
+		return fmt.Sprintf("Expected no delete, got %s", epToString(fw.lastDelete)), false
+	}
+	return "", true
+}
+
+type fakeEndpointCache struct{}
+
+func (fe *fakeEndpointCache) LookupPodName(namespacedName string) *endpoint.Endpoint {
+	return nil
+}
+
+func createCES(name, namespace string, endpoints []v2alpha1.CoreCiliumEndpoint) *v2alpha1.CiliumEndpointSlice {
+	return &v2alpha1.CiliumEndpointSlice{
+		Namespace: namespace,
+		ObjectMeta: v1.ObjectMeta{
+			Name: name,
+		},
+		Endpoints: endpoints,
+	}
+}
+
+func createEndpoint(name, namespace string, id int64) *types.CiliumEndpoint {
+	return &types.CiliumEndpoint{
+		ObjectMeta: slim_metav1.ObjectMeta{
+			Namespace: namespace,
+			Name:      name,
+		},
+		Identity:   &v2.EndpointIdentity{ID: id},
+		Encryption: &v2.EncryptionSpec{},
+	}
+}
+
+// TestCESSubscriber_CEPTransfer tests a CEP being transferred between two
+// CESes. The order of events:
+// 1. CES add event for the new CES with latest CEP state
+// 2. CES add event for the old CES with old CEP state
+// 3. CES delete event for the old CES.
+func TestCESSubscriber_CEPTransferOnStartup(t *testing.T) {
+	cepNamespace := "ns1"
+	cepName := "cep1"
+	newCEPID := int64(3)
+	oldCEPID := int64(2)
+	fakeEPWatcher := createFakeEPWatcher()
+	fakeEndpointCache := &fakeEndpointCache{}
+	cesSub := &cesSubscriber{
+		epWatcher: fakeEPWatcher,
+		epCache:   fakeEndpointCache,
+		cepMap:    newCEPToCESMap(),
+	}
+	// Add for new CES
+	cesSub.OnAdd(
+		createCES("new-ces", cepNamespace, []v2alpha1.CoreCiliumEndpoint{
+			{
+				Name:       cepName,
+				IdentityID: newCEPID,
+			},
+		}))
+	diff, ok := fakeEPWatcher.expectUpdate(endpointUpdate{
+		newEp: createEndpoint("cep1", "ns1", newCEPID),
+	})
+	if !ok {
+		t.Error(diff)
```

Comment:
Fatal here so that test exits here? If the test fails, it most likely clobber followup test steps too.

Reply:
Done